File: android/src/main/kotlin/com/example/lan_scan/handler/LanDeviceHandler.kt
package com.example.lan_scan.handler

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.example.lan_scan.networkUtils.Wireless
import com.example.lan_scan.scanner.LanDeviceScanner
import io.flutter.plugin.common.EventChannel

class LanDeviceHandler(context: Context) : EventChannel.StreamHandler {

    private val wireless = Wireless(context)
    private var lanDeviceScanner: LanDeviceScanner? = null;

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {

        if (wireless.isEnabled() && wireless.isConnectedWifi() && events != null) {
            val ipv4 = wireless.getInternalWifiIpAddress() // 获取内网IP
            val cidrPrefixLength = wireless.getInternalWifiCidrPrefixLength() // CIDR 前缀长度
            lanDeviceScanner = LanDeviceScanner(events, ipv4, cidrPrefixLength)
            lanDeviceScanner?.startScanning()
        } else {
            Handler(Looper.getMainLooper()).post {
                events?.success(null)
            }
        }
    }

    override fun onCancel(arguments: Any?) {
        lanDeviceScanner?.stopScanning()
    }
}

File: ios/Classes/handler/LanDeviceHandler.swift
//
//  LanDeviceHandler.swift
//  flutter_scan_plugin
//
//  Created by arthur on 2024/11/15.
//

import Foundation
import Flutter

class LanDeviceHandler: NSObject,FlutterStreamHandler {
    
    private var lanDeviceScanner: LanDeviceScanner?;
    
    private var wireless: Wireless = Wireless();
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        print("onListen......")
        
        if let ipv4 = wireless.getInternalWifiIpAddress(),
           let cidrPrefixLength = wireless.getInternalWifiCidrPrefixLength() {
            lanDeviceScanner = LanDeviceScanner(eventSink: events, ipv4: ipv4, cidrPrefixLength: cidrPrefixLength)
            Task{
                await lanDeviceScanner?.startScanning()
            }
            
        }else{
            DispatchQueue.main.async {
                print("no wifi")
                events(nil)
            }
        }
        
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        print("onCancel......")
        lanDeviceScanner?.stopScanning()
        return nil
    }
    
}
